pkg/cluster/px/px.go

```diff
@@ -615,6 +617,50 @@ func (ops *pxClusterOps) isScaleDownOfSharedAppsRequired(spec *apiv1alpha1.Clust
 	return opts.SharedAppsScaleDown == SharedAppsScaleDownOn, nil
 }
 
+func (ops *pxClusterOps) preFlightChecks(spec *apiv1alpha1.Cluster) error {
+	dss, err := ops.getPXDaemonsets(pxInstallTypeDocker)
+	if err != nil {
+		return err
+	}
+
+	if len(dss) > 0 {
+		return fmt.Errorf("pre-flight check failed as found: %d PX DaemonSet(s) of type docker. Only upgrading from OCI is supported", len(dss))
+	}
+
+	if len(spec.Spec.OCIMonTag) == 0 {
+		return fmt.Errorf("pre-flight check failed as new version of Portworx OCI monitor not given to upgrade API")
+	}
+
+	// check if PX is not ready in any of the nodes
+	dss, err = ops.getPXDaemonsets(pxInstallTypeOCI)
+	if err != nil {
+		return err
+	}
+
+	if len(dss) == 0 {
+		return fmt.Errorf("pre-flight check failed as it did not find any PX daemonsets for install type: %s", pxInstallTypeOCI)
+	}
+
+	for _, d := range dss {
+		err = ops.k8sOps.ValidateDaemonSet(d.Name, d.Namespace, 1*time.Minute)
+		if err != nil {
+			return fmt.Errorf("pre-flight check failed as existing Portworx DaemonSet is not ready. err: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (ops *pxClusterOps) getDaemonSetReadyTimeout() (time.Duration, error) {
+	nodes, err := ops.k8sOps.GetNodes()
+	if err != nil {
+		return 0, err
+	}
+
+	daemonsetReadyTimeout := time.Duration(len(nodes.Items)-1) * 10 * time.Minute
```

Comment:
why len(nodes) - 1 ?
also shouldn't this be just 10 minutes? Having a wait time in the order of nodes can turn out to be huge.

Reply:
- 1 because node list also contains master and no workloads run on master. 

For rolling upgrade, pods get updated one node at a time and not in parallel. So the timeout would be a function of number of nodes.

In a non-rolling upgrade scenario (just starting a new daemonset), the timeout to num nodes dependency is less, but still exists. All nodes will start pulling the docker image at the same time and all use the same upstream network pipe. So getting the same daemonset ready on 30 nodes will take more time than started on 3.